// @google-analytics/admin/build/src/v1beta/analytics_admin_service_client.js
"use strict";
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnalyticsAdminServiceClient = void 0;
const jsonProtos = require("../../protos/protos.json");
/**
 * Client JSON configuration object, loaded from
 * `src/v1beta/analytics_admin_service_client_config.json`.
 * This file defines retry strategy and timeouts for all API methods in this library.
 */
const gapicConfig = require("./analytics_admin_service_client_config.json");
const version = require('../../../package.json').version;
/**
 *  Service Interface for the Analytics Admin API (GA4).
 * @class
 * @memberof v1beta
 */
class AnalyticsAdminServiceClient {
    /**
     * Construct an instance of AnalyticsAdminServiceClient.
     *
     * @param {object} [options] - The configuration object.
     * The options accepted by the constructor are described in detail
     * in [this document](https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#creating-the-client-instance).
     * The common options are:
     * @param {object} [options.credentials] - Credentials object.
     * @param {string} [options.credentials.client_email]
     * @param {string} [options.credentials.private_key]
     * @param {string} [options.email] - Account email address. Required when
     *     using a .pem or .p12 keyFilename.
     * @param {string} [options.keyFilename] - Full path to the a .json, .pem, or
     *     .p12 key downloaded from the Google Developers Console. If you provide
     *     a path to a JSON file, the projectId option below is not necessary.
     *     NOTE: .pem and .p12 require you to specify options.email as well.
     * @param {number} [options.port] - The port on which to connect to
     *     the remote host.
     * @param {string} [options.projectId] - The project ID from the Google
     *     Developer's Console, e.g. 'grape-spaceship-123'. We will also check
     *     the environment variable GCLOUD_PROJECT for your project ID. If your
     *     app is running in an environment which supports
     *     {@link https://developers.google.com/identity/protocols/application-default-credentials Application Default Credentials},
     *     your project ID will be detected automatically.
     * @param {string} [options.apiEndpoint] - The domain name of the
     *     API remote host.
     * @param {gax.ClientConfig} [options.clientConfig] - Client configuration override.
     *     Follows the structure of {@link gapicConfig}.
     * @param {boolean} [options.fallback] - Use HTTP/1.1 REST mode.
     *     For more information, please check the
     *     {@link https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#http11-rest-api-mode documentation}.
     * @param {gax} [gaxInstance]: loaded instance of `google-gax`. Useful if you
     *     need to avoid loading the default gRPC version and want to use the fallback
     *     HTTP implementation. Load only fallback version and pass it to the constructor:
     *     ```
     *     const gax = require('google-gax/build/src/fallback'); // avoids loading google-gax with gRPC
     *     const client = new AnalyticsAdminServiceClient({fallback: true}, gax);
     *     ```
     */
    constructor(opts, gaxInstance) {
        var _a, _b, _c, _d, _e;
        this._terminated = false;
        this.descriptors = {
            page: {},
            stream: {},
            longrunning: {},
            batching: {},
        };
        // Ensure that options include all the required fields.
        const staticMembers = this
            .constructor;
        if ((opts === null || opts === void 0 ? void 0 : opts.universe_domain) &&
            (opts === null || opts === void 0 ? void 0 : opts.universeDomain) &&
            (opts === null || opts === void 0 ? void 0 : opts.universe_domain) !== (opts === null || opts === void 0 ? void 0 : opts.universeDomain)) {
            throw new Error('Please set either universe_domain or universeDomain, but not both.');
        }
        const universeDomainEnvVar = typeof process === 'object' && typeof process.env === 'object'
            ? process.env['GOOGLE_CLOUD_UNIVERSE_DOMAIN']
            : undefined;
        this._universeDomain =
            (_c = (_b = (_a = opts === null || opts === void 0 ? void 0 : opts.universeDomain) !== null && _a !== void 0 ? _a : opts === null || opts === void 0 ? void 0 : opts.universe_domain) !== null && _b !== void 0 ? _b : universeDomainEnvVar) !== null && _c !== void 0 ? _c : 'googleapis.com';
        this._servicePath = 'analyticsadmin.' + this._universeDomain;
        const servicePath = (opts === null || opts === void 0 ? void 0 : opts.servicePath) || (opts === null || opts === void 0 ? void 0 : opts.apiEndpoint) || this._servicePath;
        this._providedCustomServicePath = !!((opts === null || opts === void 0 ? void 0 : opts.servicePath) || (opts === null || opts === void 0 ? void 0 : opts.apiEndpoint));
        const port = (opts === null || opts === void 0 ? void 0 : opts.port) || staticMembers.port;
        const clientConfig = (_d = opts === null || opts === void 0 ? void 0 : opts.clientConfig) !== null && _d !== void 0 ? _d : {};
        const fallback = (_e = opts === null || opts === void 0 ? void 0 : opts.fallback) !== null && _e !== void 0 ? _e : (typeof window !== 'undefined' && typeof (window === null || window === void 0 ? void 0 : window.fetch) === 'function');
        opts = Object.assign({ servicePath, port, clientConfig, fallback }, opts);
        // Request numeric enum values if REST transport is used.
        opts.numericEnums = true;
        // If scopes are unset in options and we're connecting to a non-default endpoint, set scopes just in case.
        if (servicePath !== this._servicePath && !('scopes' in opts)) {
            opts['scopes'] = staticMembers.scopes;
        }
        // Load google-gax module synchronously if needed
        if (!gaxInstance) {
            gaxInstance = require('google-gax');
        }
        // Choose either gRPC or proto-over-HTTP implementation of google-gax.
        this._gaxModule = opts.fallback ? gaxInstance.fallback : gaxInstance;
        // Create a `gaxGrpc` object, with any grpc-specific options sent to the client.
        this._gaxGrpc = new this._gaxModule.GrpcClient(opts);
        // Save options to use in initialize() method.
        this._opts = opts;
        // Save the auth object to the client, for use by other methods.
        this.auth = this._gaxGrpc.auth;
        // Set useJWTAccessWithScope on the auth object.
        this.auth.useJWTAccessWithScope = true;
        // Set defaultServicePath on the auth object.
        this.auth.defaultServicePath = this._servicePath;
        // Set the default scopes in auth client if needed.
        if (servicePath === this._servicePath) {
            this.auth.defaultScopes = staticMembers.scopes;
        }
        // Determine the client header string.
        const clientHeader = [`gax/${this._gaxModule.version}`, `gapic/${version}`];
        if (typeof process === 'object' && 'versions' in process) {
            clientHeader.push(`gl-node/${process.versions.node}`);
        }
        else {
            clientHeader.push(`gl-web/${this._gaxModule.version}`);
        }
        if (!opts.fallback) {
            clientHeader.push(`grpc/${this._gaxGrpc.grpcVersion}`);
        }
        else {
            clientHeader.push(`rest/${this._gaxGrpc.grpcVersion}`);
        }
        if (opts.libName && opts.libVersion) {
            clientHeader.push(`${opts.libName}/${opts.libVersion}`);
        }
        // Load the applicable protos.
        this._protos = this._gaxGrpc.loadProtoJSON(jsonProtos);
        // This API contains "path templates"; forward-slash-separated
        // identifiers to uniquely identify resources within the API.
        // Create useful helper objects for these.
        this.pathTemplates = {
            accountPathTemplate: new this._gaxModule.PathTemplate('accounts/{account}'),
            accountSummaryPathTemplate: new this._gaxModule.PathTemplate('accountSummaries/{account_summary}'),
            conversionEventPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/conversionEvents/{conversion_event}'),
            customDimensionPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/customDimensions/{custom_dimension}'),
            customMetricPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/customMetrics/{custom_metric}'),
            dataRetentionSettingsPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/dataRetentionSettings'),
            dataSharingSettingsPathTemplate: new this._gaxModule.PathTemplate('accounts/{account}/dataSharingSettings'),
            dataStreamPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/dataStreams/{data_stream}'),
            firebaseLinkPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/firebaseLinks/{firebase_link}'),
            googleAdsLinkPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/googleAdsLinks/{google_ads_link}'),
            keyEventPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/keyEvents/{key_event}'),
            measurementProtocolSecretPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/dataStreams/{data_stream}/measurementProtocolSecrets/{measurement_protocol_secret}'),
            propertyPathTemplate: new this._gaxModule.PathTemplate('properties/{property}'),
        };
        // Some of the methods on this service return "paged" results,
        // (e.g. 50 results at a time, with tokens to get subsequent
        // pages). Denote the keys used for pagination and results.
        this.descriptors.page = {
            listAccounts: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'accounts'),
            listAccountSummaries: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'accountSummaries'),
            listProperties: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'properties'),
            listFirebaseLinks: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'firebaseLinks'),
            listGoogleAdsLinks: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'googleAdsLinks'),
            listMeasurementProtocolSecrets: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'measurementProtocolSecrets'),
            searchChangeHistoryEvents: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'changeHistoryEvents'),
            listConversionEvents: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'conversionEvents'),
            listKeyEvents: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'keyEvents'),
            listCustomDimensions: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'customDimensions'),
            listCustomMetrics: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'customMetrics'),
            listDataStreams: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'dataStreams'),
        };
        // Put together the default options sent with requests.
        this._defaults = this._gaxGrpc.constructSettings('google.analytics.admin.v1beta.AnalyticsAdminService', gapicConfig, opts.clientConfig || {}, { 'x-goog-api-client': clientHeader.join(' ') });
        // Set up a dictionary of "inner API calls"; the core implementation
        // of calling the API is handled in `google-gax`, with this code
        // merely providing the destination and request information.
        this.innerApiCalls = {};
        // Add a warn function to the client constructor so it can be easily tested.
        this.warn = this._gaxModule.warn;
    }
    /**
     * Initialize the client.
     * Performs asynchronous operations (such as authentication) and prepares the client.
     * This function will be called automatically when any class method is called for the
     * first time, but if you need to initialize it before calling an actual method,
     * feel free to call initialize() directly.
     *
     * You can await on this method if you want to make sure the client is initialized.
     *
     * @returns {Promise} A promise that resolves to an authenticated service stub.
     */
    initialize() {
        // If the client stub promise is already initialized, return immediately.
        if (this.analyticsAdminServiceStub) {
            return this.analyticsAdminServiceStub;
        }
        // Put together the "service stub" for
        // google.analytics.admin.v1beta.AnalyticsAdminService.
        this.analyticsAdminServiceStub = this._gaxGrpc.createStub(this._opts.fallback
            ? this._protos.lookupService('google.analytics.admin.v1beta.AnalyticsAdminService')
            : // eslint-disable-next-line @typescript-eslint/no-explicit-any
                this._protos.google.analytics.admin.v1beta
                    .AnalyticsAdminService, this._opts, this._providedCustomServicePath);
        // Iterate over each of the methods that the service provides
        // and create an API call method for each.
        const analyticsAdminServiceStubMethods = [
            'getAccount',
            'listAccounts',
            'deleteAccount',
            'updateAccount',
            'provisionAccountTicket',
            'listAccountSummaries',
            'getProperty',
            'listProperties',
            'createProperty',
            'deleteProperty',
            'updateProperty',
            'createFirebaseLink',
            'deleteFirebaseLink',
            'listFirebaseLinks',
            'createGoogleAdsLink',
            'updateGoogleAdsLink',
            'deleteGoogleAdsLink',
            'listGoogleAdsLinks',
            'getDataSharingSettings',
            'getMeasurementProtocolSecret',
            'listMeasurementProtocolSecrets',
            'createMeasurementProtocolSecret',
            'deleteMeasurementProtocolSecret',
            'updateMeasurementProtocolSecret',
            'acknowledgeUserDataCollection',
            'searchChangeHistoryEvents',
            'createConversionEvent',
            'updateConversionEvent',
            'getConversionEvent',
            'deleteConversionEvent',
            'listConversionEvents',
            'createKeyEvent',
            'updateKeyEvent',
            'getKeyEvent',
            'deleteKeyEvent',
            'listKeyEvents',
            'createCustomDimension',
            'updateCustomDimension',
            'listCustomDimensions',
            'archiveCustomDimension',
            'getCustomDimension',
            'createCustomMetric',
            'updateCustomMetric',
            'listCustomMetrics',
            'archiveCustomMetric',
            'getCustomMetric',
            'getDataRetentionSettings',
            'updateDataRetentionSettings',
            'createDataStream',
            'deleteDataStream',
            'updateDataStream',
            'listDataStreams',
            'getDataStream',
            'runAccessReport',
        ];
        for (const methodName of analyticsAdminServiceStubMethods) {
            const callPromise = this.analyticsAdminServiceStub.then(stub => (...args) => {
                if (this._terminated) {
                    return Promise.reject('The client has already been closed.');
                }
                const func = stub[methodName];
                return func.apply(stub, args);
            }, (err) => () => {
                throw err;
            });
            const descriptor = this.descriptors.page[methodName] || undefined;
            const apiCall = this._gaxModule.createApiCall(callPromise, this._defaults[methodName], descriptor, this._opts.fallback);
            this.innerApiCalls[methodName] = apiCall;
        }
        return this.analyticsAdminServiceStub;
    }
    /**
     * The DNS address for this API service.
     * @deprecated Use the apiEndpoint method of the client instance.
     * @returns {string} The DNS address for this service.
     */
    static get servicePath() {
        if (typeof process === 'object' &&
            typeof process.emitWarning === 'function') {
            process.emitWarning('Static servicePath is deprecated, please use the instance method instead.', 'DeprecationWarning');
        }
        return 'analyticsadmin.googleapis.com';
    }
    /**
     * The DNS address for this API service - same as servicePath.
     * @deprecated Use the apiEndpoint method of the client instance.
     * @returns {string} The DNS address for this service.
     */
    static get apiEndpoint() {
        if (typeof process === 'object' &&
            typeof process.emitWarning === 'function') {
            process.emitWarning('Static apiEndpoint is deprecated, please use the instance method instead.', 'DeprecationWarning');
        }
        return 'analyticsadmin.googleapis.com';
    }
    /**
     * The DNS address for this API service.
     * @returns {string} The DNS address for this service.
     */
    get apiEndpoint() {
        return this._servicePath;
    }
    get universeDomain() {
        return this._universeDomain;
    }
    /**
     * The port for this API service.
     * @returns {number} The default port for this service.
     */
    static get port() {
        return 443;
    }
    /**
     * The scopes needed to make gRPC calls for every method defined
     * in this service.
     * @returns {string[]} List of default scopes.
     */
    static get scopes() {
        return [
            'https://www.googleapis.com/auth/analytics.edit',
            'https://www.googleapis.com/auth/analytics.readonly',
        ];
    }
    /**
     * Return the project ID used by this class.
     * @returns {Promise} A promise that resolves to string containing the project ID.
     */
    getProjectId(callback) {
        if (callback) {
            this.auth.getProjectId(callback);
            return;
        }
        return this.auth.getProjectId();
    }
    getAccount(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.getAccount(request, options, callback);
    }
    deleteAccount(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.deleteAccount(request, options, callback);
    }
    updateAccount(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'account.name': (_a = request.account.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.updateAccount(request, options, callback);
    }
    provisionAccountTicket(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        this.initialize();
        return this.innerApiCalls.provisionAccountTicket(request, options, callback);
    }
    getProperty(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.getProperty(request, options, callback);
    }
    createProperty(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        this.initialize();
        return this.innerApiCalls.createProperty(request, options, callback);
    }
    deleteProperty(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.deleteProperty(request, options, callback);
    }
    updateProperty(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'property.name': (_a = request.property.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.updateProperty(request, options, callback);
    }
    createFirebaseLink(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.createFirebaseLink(request, options, callback);
    }
    deleteFirebaseLink(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.deleteFirebaseLink(request, options, callback);
    }
    createGoogleAdsLink(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.createGoogleAdsLink(request, options, callback);
    }
    updateGoogleAdsLink(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'google_ads_link.name': (_a = request.googleAdsLink.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.updateGoogleAdsLink(request, options, callback);
    }
    deleteGoogleAdsLink(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.deleteGoogleAdsLink(request, options, callback);
    }
    getDataSharingSettings(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.getDataSharingSettings(request, options, callback);
    }
    getMeasurementProtocolSecret(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.getMeasurementProtocolSecret(request, options, callback);
    }
    createMeasurementProtocolSecret(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.createMeasurementProtocolSecret(request, options, callback);
    }
    deleteMeasurementProtocolSecret(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.deleteMeasurementProtocolSecret(request, options, callback);
    }
    updateMeasurementProtocolSecret(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'measurement_protocol_secret.name': (_a = request.measurementProtocolSecret.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.updateMeasurementProtocolSecret(request, options, callback);
    }
    acknowledgeUserDataCollection(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                property: (_a = request.property) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.acknowledgeUserDataCollection(request, options, callback);
    }
    createConversionEvent(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        this.warn('DEP$AnalyticsAdminService-$CreateConversionEvent', 'CreateConversionEvent is deprecated and may be removed in a future version.', 'DeprecationWarning');
        return this.innerApiCalls.createConversionEvent(request, options, callback);
    }
    updateConversionEvent(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'conversion_event.name': (_a = request.conversionEvent.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        this.warn('DEP$AnalyticsAdminService-$UpdateConversionEvent', 'UpdateConversionEvent is deprecated and may be removed in a future version.', 'DeprecationWarning');
        return this.innerApiCalls.updateConversionEvent(request, options, callback);
    }
    getConversionEvent(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        this.warn('DEP$AnalyticsAdminService-$GetConversionEvent', 'GetConversionEvent is deprecated and may be removed in a future version.', 'DeprecationWarning');
        return this.innerApiCalls.getConversionEvent(request, options, callback);
    }
    deleteConversionEvent(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        this.warn('DEP$AnalyticsAdminService-$DeleteConversionEvent', 'DeleteConversionEvent is deprecated and may be removed in a future version.', 'DeprecationWarning');
        return this.innerApiCalls.deleteConversionEvent(request, options, callback);
    }
    createKeyEvent(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.createKeyEvent(request, options, callback);
    }
    updateKeyEvent(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'key_event.name': (_a = request.keyEvent.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.updateKeyEvent(request, options, callback);
    }
    getKeyEvent(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.getKeyEvent(request, options, callback);
    }
    deleteKeyEvent(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.deleteKeyEvent(request, options, callback);
    }
    createCustomDimension(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.createCustomDimension(request, options, callback);
    }
    updateCustomDimension(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'custom_dimension.name': (_a = request.customDimension.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.updateCustomDimension(request, options, callback);
    }
    archiveCustomDimension(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.archiveCustomDimension(request, options, callback);
    }
    getCustomDimension(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.getCustomDimension(request, options, callback);
    }
    createCustomMetric(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.createCustomMetric(request, options, callback);
    }
    updateCustomMetric(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'custom_metric.name': (_a = request.customMetric.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.updateCustomMetric(request, options, callback);
    }
    archiveCustomMetric(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.archiveCustomMetric(request, options, callback);
    }
    getCustomMetric(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.getCustomMetric(request, options, callback);
    }
    getDataRetentionSettings(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.getDataRetentionSettings(request, options, callback);
    }
    updateDataRetentionSettings(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'data_retention_settings.name': (_a = request.dataRetentionSettings.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.updateDataRetentionSettings(request, options, callback);
    }
    createDataStream(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.createDataStream(request, options, callback);
    }
    deleteDataStream(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.deleteDataStream(request, options, callback);
    }
    updateDataStream(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                'data_stream.name': (_a = request.dataStream.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.updateDataStream(request, options, callback);
    }
    getDataStream(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                name: (_a = request.name) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.getDataStream(request, options, callback);
    }
    runAccessReport(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                entity: (_a = request.entity) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.runAccessReport(request, options, callback);
    }
    listAccounts(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        this.initialize();
        return this.innerApiCalls.listAccounts(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {number} request.pageSize
     *   The maximum number of resources to return. The service may return
     *   fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListAccounts` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListAccounts` must
     *   match the call that provided the page token.
     * @param {boolean} request.showDeleted
     *   Whether to include soft-deleted (ie: "trashed") Accounts in the
     *   results. Accounts can be inspected to determine whether they are deleted or
     *   not.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.Account|Account} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listAccountsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listAccountsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const defaultCallSettings = this._defaults['listAccounts'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listAccounts.createStream(this.innerApiCalls.listAccounts, request, callSettings);
    }
    /**
     * Equivalent to `listAccounts`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {number} request.pageSize
     *   The maximum number of resources to return. The service may return
     *   fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListAccounts` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListAccounts` must
     *   match the call that provided the page token.
     * @param {boolean} request.showDeleted
     *   Whether to include soft-deleted (ie: "trashed") Accounts in the
     *   results. Accounts can be inspected to determine whether they are deleted or
     *   not.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.Account|Account}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_accounts.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListAccounts_async
     */
    listAccountsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const defaultCallSettings = this._defaults['listAccounts'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listAccounts.asyncIterate(this.innerApiCalls['listAccounts'], request, callSettings);
    }
    listAccountSummaries(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        this.initialize();
        return this.innerApiCalls.listAccountSummaries(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {number} request.pageSize
     *   The maximum number of AccountSummary resources to return. The service may
     *   return fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListAccountSummaries` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListAccountSummaries`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.AccountSummary|AccountSummary} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listAccountSummariesAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listAccountSummariesStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const defaultCallSettings = this._defaults['listAccountSummaries'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listAccountSummaries.createStream(this.innerApiCalls.listAccountSummaries, request, callSettings);
    }
    /**
     * Equivalent to `listAccountSummaries`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {number} request.pageSize
     *   The maximum number of AccountSummary resources to return. The service may
     *   return fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListAccountSummaries` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListAccountSummaries`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.AccountSummary|AccountSummary}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_account_summaries.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListAccountSummaries_async
     */
    listAccountSummariesAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const defaultCallSettings = this._defaults['listAccountSummaries'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listAccountSummaries.asyncIterate(this.innerApiCalls['listAccountSummaries'], request, callSettings);
    }
    listProperties(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        this.initialize();
        return this.innerApiCalls.listProperties(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.filter
     *   Required. An expression for filtering the results of the request.
     *   Fields eligible for filtering are:
     *   `parent:`(The resource name of the parent account/property) or
     *   `ancestor:`(The resource name of the parent account) or
     *   `firebase_project:`(The id or number of the linked firebase project).
     *   Some examples of filters:
     *
     *   ```
     *   | Filter                      | Description                               |
     *   |-----------------------------|-------------------------------------------|
     *   | parent:accounts/123         | The account with account id: 123.       |
     *   | parent:properties/123       | The property with property id: 123.       |
     *   | ancestor:accounts/123       | The account with account id: 123.         |
     *   | firebase_project:project-id | The firebase project with id: project-id. |
     *   | firebase_project:123        | The firebase project with number: 123.    |
     *   ```
     * @param {number} request.pageSize
     *   The maximum number of resources to return. The service may return
     *   fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListProperties` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListProperties` must
     *   match the call that provided the page token.
     * @param {boolean} request.showDeleted
     *   Whether to include soft-deleted (ie: "trashed") Properties in the
     *   results. Properties can be inspected to determine whether they are deleted
     *   or not.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.Property|Property} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listPropertiesAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listPropertiesStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const defaultCallSettings = this._defaults['listProperties'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listProperties.createStream(this.innerApiCalls.listProperties, request, callSettings);
    }
    /**
     * Equivalent to `listProperties`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.filter
     *   Required. An expression for filtering the results of the request.
     *   Fields eligible for filtering are:
     *   `parent:`(The resource name of the parent account/property) or
     *   `ancestor:`(The resource name of the parent account) or
     *   `firebase_project:`(The id or number of the linked firebase project).
     *   Some examples of filters:
     *
     *   ```
     *   | Filter                      | Description                               |
     *   |-----------------------------|-------------------------------------------|
     *   | parent:accounts/123         | The account with account id: 123.       |
     *   | parent:properties/123       | The property with property id: 123.       |
     *   | ancestor:accounts/123       | The account with account id: 123.         |
     *   | firebase_project:project-id | The firebase project with id: project-id. |
     *   | firebase_project:123        | The firebase project with number: 123.    |
     *   ```
     * @param {number} request.pageSize
     *   The maximum number of resources to return. The service may return
     *   fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListProperties` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListProperties` must
     *   match the call that provided the page token.
     * @param {boolean} request.showDeleted
     *   Whether to include soft-deleted (ie: "trashed") Properties in the
     *   results. Properties can be inspected to determine whether they are deleted
     *   or not.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.Property|Property}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_properties.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListProperties_async
     */
    listPropertiesAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const defaultCallSettings = this._defaults['listProperties'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listProperties.asyncIterate(this.innerApiCalls['listProperties'], request, callSettings);
    }
    listFirebaseLinks(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.listFirebaseLinks(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Format: properties/{property_id}
     *
     *   Example: `properties/1234`
     * @param {number} request.pageSize
     *   The maximum number of resources to return. The service may return
     *   fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListFirebaseLinks` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListFirebaseLinks` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.FirebaseLink|FirebaseLink} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listFirebaseLinksAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listFirebaseLinksStream(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listFirebaseLinks'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listFirebaseLinks.createStream(this.innerApiCalls.listFirebaseLinks, request, callSettings);
    }
    /**
     * Equivalent to `listFirebaseLinks`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Format: properties/{property_id}
     *
     *   Example: `properties/1234`
     * @param {number} request.pageSize
     *   The maximum number of resources to return. The service may return
     *   fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListFirebaseLinks` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListFirebaseLinks` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.FirebaseLink|FirebaseLink}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_firebase_links.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListFirebaseLinks_async
     */
    listFirebaseLinksAsync(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listFirebaseLinks'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listFirebaseLinks.asyncIterate(this.innerApiCalls['listFirebaseLinks'], request, callSettings);
    }
    listGoogleAdsLinks(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.listGoogleAdsLinks(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListGoogleAdsLinks` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListGoogleAdsLinks` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.GoogleAdsLink|GoogleAdsLink} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listGoogleAdsLinksAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listGoogleAdsLinksStream(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listGoogleAdsLinks'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listGoogleAdsLinks.createStream(this.innerApiCalls.listGoogleAdsLinks, request, callSettings);
    }
    /**
     * Equivalent to `listGoogleAdsLinks`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListGoogleAdsLinks` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListGoogleAdsLinks` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.GoogleAdsLink|GoogleAdsLink}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_google_ads_links.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListGoogleAdsLinks_async
     */
    listGoogleAdsLinksAsync(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listGoogleAdsLinks'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listGoogleAdsLinks.asyncIterate(this.innerApiCalls['listGoogleAdsLinks'], request, callSettings);
    }
    listMeasurementProtocolSecrets(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.listMeasurementProtocolSecrets(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The resource name of the parent stream.
     *   Format:
     *   properties/{property}/dataStreams/{dataStream}/measurementProtocolSecrets
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 10 resources will be returned.
     *   The maximum value is 10. Higher values will be coerced to the maximum.
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListMeasurementProtocolSecrets`
     *   call. Provide this to retrieve the subsequent page. When paginating, all
     *   other parameters provided to `ListMeasurementProtocolSecrets` must match
     *   the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.MeasurementProtocolSecret|MeasurementProtocolSecret} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listMeasurementProtocolSecretsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listMeasurementProtocolSecretsStream(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listMeasurementProtocolSecrets'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listMeasurementProtocolSecrets.createStream(this.innerApiCalls.listMeasurementProtocolSecrets, request, callSettings);
    }
    /**
     * Equivalent to `listMeasurementProtocolSecrets`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The resource name of the parent stream.
     *   Format:
     *   properties/{property}/dataStreams/{dataStream}/measurementProtocolSecrets
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 10 resources will be returned.
     *   The maximum value is 10. Higher values will be coerced to the maximum.
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListMeasurementProtocolSecrets`
     *   call. Provide this to retrieve the subsequent page. When paginating, all
     *   other parameters provided to `ListMeasurementProtocolSecrets` must match
     *   the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.MeasurementProtocolSecret|MeasurementProtocolSecret}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_measurement_protocol_secrets.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListMeasurementProtocolSecrets_async
     */
    listMeasurementProtocolSecretsAsync(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listMeasurementProtocolSecrets'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listMeasurementProtocolSecrets.asyncIterate(this.innerApiCalls['listMeasurementProtocolSecrets'], request, callSettings);
    }
    searchChangeHistoryEvents(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                account: (_a = request.account) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.searchChangeHistoryEvents(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.account
     *   Required. The account resource for which to return change history
     *   resources. Format: accounts/{account}
     *
     *   Example: `accounts/100`
     * @param {string} [request.property]
     *   Optional. Resource name for a child property. If set, only return changes
     *   made to this property or its child resources.
     *   Format: properties/{propertyId}
     *
     *   Example: `properties/100`
     * @param {number[]} [request.resourceType]
     *   Optional. If set, only return changes if they are for a resource that
     *   matches at least one of these types.
     * @param {number[]} [request.action]
     *   Optional. If set, only return changes that match one or more of these types
     *   of actions.
     * @param {string[]} [request.actorEmail]
     *   Optional. If set, only return changes if they are made by a user in this
     *   list.
     * @param {google.protobuf.Timestamp} [request.earliestChangeTime]
     *   Optional. If set, only return changes made after this time (inclusive).
     * @param {google.protobuf.Timestamp} [request.latestChangeTime]
     *   Optional. If set, only return changes made before this time (inclusive).
     * @param {number} [request.pageSize]
     *   Optional. The maximum number of ChangeHistoryEvent items to return.
     *   The service may return fewer than this value, even if there are additional
     *   pages. If unspecified, at most 50 items will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} [request.pageToken]
     *   Optional. A page token, received from a previous
     *   `SearchChangeHistoryEvents` call. Provide this to retrieve the subsequent
     *   page. When paginating, all other parameters provided to
     *   `SearchChangeHistoryEvents` must match the call that provided the page
     *   token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.ChangeHistoryEvent|ChangeHistoryEvent} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `searchChangeHistoryEventsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    searchChangeHistoryEventsStream(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                account: (_a = request.account) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['searchChangeHistoryEvents'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.searchChangeHistoryEvents.createStream(this.innerApiCalls.searchChangeHistoryEvents, request, callSettings);
    }
    /**
     * Equivalent to `searchChangeHistoryEvents`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.account
     *   Required. The account resource for which to return change history
     *   resources. Format: accounts/{account}
     *
     *   Example: `accounts/100`
     * @param {string} [request.property]
     *   Optional. Resource name for a child property. If set, only return changes
     *   made to this property or its child resources.
     *   Format: properties/{propertyId}
     *
     *   Example: `properties/100`
     * @param {number[]} [request.resourceType]
     *   Optional. If set, only return changes if they are for a resource that
     *   matches at least one of these types.
     * @param {number[]} [request.action]
     *   Optional. If set, only return changes that match one or more of these types
     *   of actions.
     * @param {string[]} [request.actorEmail]
     *   Optional. If set, only return changes if they are made by a user in this
     *   list.
     * @param {google.protobuf.Timestamp} [request.earliestChangeTime]
     *   Optional. If set, only return changes made after this time (inclusive).
     * @param {google.protobuf.Timestamp} [request.latestChangeTime]
     *   Optional. If set, only return changes made before this time (inclusive).
     * @param {number} [request.pageSize]
     *   Optional. The maximum number of ChangeHistoryEvent items to return.
     *   The service may return fewer than this value, even if there are additional
     *   pages. If unspecified, at most 50 items will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} [request.pageToken]
     *   Optional. A page token, received from a previous
     *   `SearchChangeHistoryEvents` call. Provide this to retrieve the subsequent
     *   page. When paginating, all other parameters provided to
     *   `SearchChangeHistoryEvents` must match the call that provided the page
     *   token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.ChangeHistoryEvent|ChangeHistoryEvent}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.search_change_history_events.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_SearchChangeHistoryEvents_async
     */
    searchChangeHistoryEventsAsync(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                account: (_a = request.account) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['searchChangeHistoryEvents'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.searchChangeHistoryEvents.asyncIterate(this.innerApiCalls['searchChangeHistoryEvents'], request, callSettings);
    }
    listConversionEvents(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        this.warn('DEP$AnalyticsAdminService-$ListConversionEvents', 'ListConversionEvents is deprecated and may be removed in a future version.', 'DeprecationWarning');
        return this.innerApiCalls.listConversionEvents(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The resource name of the parent property.
     *   Example: 'properties/123'
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListConversionEvents` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListConversionEvents`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.ConversionEvent|ConversionEvent} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listConversionEventsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @deprecated ListConversionEvents is deprecated and may be removed in a future version.
     */
    listConversionEventsStream(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listConversionEvents'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        this.warn('DEP$AnalyticsAdminService-$ListConversionEvents', 'ListConversionEvents is deprecated and may be removed in a future version.', 'DeprecationWarning');
        return this.descriptors.page.listConversionEvents.createStream(this.innerApiCalls.listConversionEvents, request, callSettings);
    }
    /**
     * Equivalent to `listConversionEvents`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The resource name of the parent property.
     *   Example: 'properties/123'
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListConversionEvents` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListConversionEvents`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.ConversionEvent|ConversionEvent}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_conversion_events.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListConversionEvents_async
     * @deprecated ListConversionEvents is deprecated and may be removed in a future version.
     */
    listConversionEventsAsync(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listConversionEvents'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        this.warn('DEP$AnalyticsAdminService-$ListConversionEvents', 'ListConversionEvents is deprecated and may be removed in a future version.', 'DeprecationWarning');
        return this.descriptors.page.listConversionEvents.asyncIterate(this.innerApiCalls['listConversionEvents'], request, callSettings);
    }
    listKeyEvents(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.listKeyEvents(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The resource name of the parent property.
     *   Example: 'properties/123'
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListKeyEvents` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListKeyEvents`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.KeyEvent|KeyEvent} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listKeyEventsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listKeyEventsStream(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listKeyEvents'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listKeyEvents.createStream(this.innerApiCalls.listKeyEvents, request, callSettings);
    }
    /**
     * Equivalent to `listKeyEvents`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The resource name of the parent property.
     *   Example: 'properties/123'
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListKeyEvents` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListKeyEvents`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.KeyEvent|KeyEvent}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_key_events.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListKeyEvents_async
     */
    listKeyEventsAsync(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listKeyEvents'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listKeyEvents.asyncIterate(this.innerApiCalls['listKeyEvents'], request, callSettings);
    }
    listCustomDimensions(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.listCustomDimensions(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListCustomDimensions` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListCustomDimensions`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.CustomDimension|CustomDimension} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listCustomDimensionsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listCustomDimensionsStream(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listCustomDimensions'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listCustomDimensions.createStream(this.innerApiCalls.listCustomDimensions, request, callSettings);
    }
    /**
     * Equivalent to `listCustomDimensions`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListCustomDimensions` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListCustomDimensions`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.CustomDimension|CustomDimension}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_custom_dimensions.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListCustomDimensions_async
     */
    listCustomDimensionsAsync(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listCustomDimensions'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listCustomDimensions.asyncIterate(this.innerApiCalls['listCustomDimensions'], request, callSettings);
    }
    listCustomMetrics(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.listCustomMetrics(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListCustomMetrics` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListCustomMetrics` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.CustomMetric|CustomMetric} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listCustomMetricsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listCustomMetricsStream(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listCustomMetrics'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listCustomMetrics.createStream(this.innerApiCalls.listCustomMetrics, request, callSettings);
    }
    /**
     * Equivalent to `listCustomMetrics`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListCustomMetrics` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListCustomMetrics` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.CustomMetric|CustomMetric}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_custom_metrics.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListCustomMetrics_async
     */
    listCustomMetricsAsync(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listCustomMetrics'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listCustomMetrics.asyncIterate(this.innerApiCalls['listCustomMetrics'], request, callSettings);
    }
    listDataStreams(request, optionsOrCallback, callback) {
        var _a;
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        this.initialize();
        return this.innerApiCalls.listDataStreams(request, options, callback);
    }
    /**
     * Equivalent to `method.name.toCamelCase()`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListDataStreams` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListDataStreams` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.DataStream|DataStream} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listDataStreamsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listDataStreamsStream(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listDataStreams'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listDataStreams.createStream(this.innerApiCalls.listDataStreams, request, callSettings);
    }
    /**
     * Equivalent to `listDataStreams`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListDataStreams` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListDataStreams` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.DataStream|DataStream}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_data_streams.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListDataStreams_async
     */
    listDataStreamsAsync(request, options) {
        var _a;
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] =
            this._gaxModule.routingHeader.fromParams({
                parent: (_a = request.parent) !== null && _a !== void 0 ? _a : '',
            });
        const defaultCallSettings = this._defaults['listDataStreams'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize();
        return this.descriptors.page.listDataStreams.asyncIterate(this.innerApiCalls['listDataStreams'], request, callSettings);
    }
    // --------------------
    // -- Path templates --
    // --------------------
    /**
     * Return a fully-qualified account resource name string.
     *
     * @param {string} account
     * @returns {string} Resource name string.
     */
    accountPath(account) {
        return this.pathTemplates.accountPathTemplate.render({
            account: account,
        });
    }
    /**
     * Parse the account from Account resource.
     *
     * @param {string} accountName
     *   A fully-qualified path representing Account resource.
     * @returns {string} A string representing the account.
     */
    matchAccountFromAccountName(accountName) {
        return this.pathTemplates.accountPathTemplate.match(accountName).account;
    }
    /**
     * Return a fully-qualified accountSummary resource name string.
     *
     * @param {string} account_summary
     * @returns {string} Resource name string.
     */
    accountSummaryPath(accountSummary) {
        return this.pathTemplates.accountSummaryPathTemplate.render({
            account_summary: accountSummary,
        });
    }
    /**
     * Parse the account_summary from AccountSummary resource.
     *
     * @param {string} accountSummaryName
     *   A fully-qualified path representing AccountSummary resource.
     * @returns {string} A string representing the account_summary.
     */
    matchAccountSummaryFromAccountSummaryName(accountSummaryName) {
        return this.pathTemplates.accountSummaryPathTemplate.match(accountSummaryName).account_summary;
    }
    /**
     * Return a fully-qualified conversionEvent resource name string.
     *
     * @param {string} property
     * @param {string} conversion_event
     * @returns {string} Resource name string.
     */
    conversionEventPath(property, conversionEvent) {
        return this.pathTemplates.conversionEventPathTemplate.render({
            property: property,
            conversion_event: conversionEvent,
        });
    }
    /**
     * Parse the property from ConversionEvent resource.
     *
     * @param {string} conversionEventName
     *   A fully-qualified path representing ConversionEvent resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromConversionEventName(conversionEventName) {
        return this.pathTemplates.conversionEventPathTemplate.match(conversionEventName).property;
    }
    /**
     * Parse the conversion_event from ConversionEvent resource.
     *
     * @param {string} conversionEventName
     *   A fully-qualified path representing ConversionEvent resource.
     * @returns {string} A string representing the conversion_event.
     */
    matchConversionEventFromConversionEventName(conversionEventName) {
        return this.pathTemplates.conversionEventPathTemplate.match(conversionEventName).conversion_event;
    }
    /**
     * Return a fully-qualified customDimension resource name string.
     *
     * @param {string} property
     * @param {string} custom_dimension
     * @returns {string} Resource name string.
     */
    customDimensionPath(property, customDimension) {
        return this.pathTemplates.customDimensionPathTemplate.render({
            property: property,
            custom_dimension: customDimension,
        });
    }
    /**
     * Parse the property from CustomDimension resource.
     *
     * @param {string} customDimensionName
     *   A fully-qualified path representing CustomDimension resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromCustomDimensionName(customDimensionName) {
        return this.pathTemplates.customDimensionPathTemplate.match(customDimensionName).property;
    }
    /**
     * Parse the custom_dimension from CustomDimension resource.
     *
     * @param {string} customDimensionName
     *   A fully-qualified path representing CustomDimension resource.
     * @returns {string} A string representing the custom_dimension.
     */
    matchCustomDimensionFromCustomDimensionName(customDimensionName) {
        return this.pathTemplates.customDimensionPathTemplate.match(customDimensionName).custom_dimension;
    }
    /**
     * Return a fully-qualified customMetric resource name string.
     *
     * @param {string} property
     * @param {string} custom_metric
     * @returns {string} Resource name string.
     */
    customMetricPath(property, customMetric) {
        return this.pathTemplates.customMetricPathTemplate.render({
            property: property,
            custom_metric: customMetric,
        });
    }
    /**
     * Parse the property from CustomMetric resource.
     *
     * @param {string} customMetricName
     *   A fully-qualified path representing CustomMetric resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromCustomMetricName(customMetricName) {
        return this.pathTemplates.customMetricPathTemplate.match(customMetricName)
            .property;
    }
    /**
     * Parse the custom_metric from CustomMetric resource.
     *
     * @param {string} customMetricName
     *   A fully-qualified path representing CustomMetric resource.
     * @returns {string} A string representing the custom_metric.
     */
    matchCustomMetricFromCustomMetricName(customMetricName) {
        return this.pathTemplates.customMetricPathTemplate.match(customMetricName)
            .custom_metric;
    }
    /**
     * Return a fully-qualified dataRetentionSettings resource name string.
     *
     * @param {string} property
     * @returns {string} Resource name string.
     */
    dataRetentionSettingsPath(property) {
        return this.pathTemplates.dataRetentionSettingsPathTemplate.render({
            property: property,
        });
    }
    /**
     * Parse the property from DataRetentionSettings resource.
     *
     * @param {string} dataRetentionSettingsName
     *   A fully-qualified path representing DataRetentionSettings resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromDataRetentionSettingsName(dataRetentionSettingsName) {
        return this.pathTemplates.dataRetentionSettingsPathTemplate.match(dataRetentionSettingsName).property;
    }
    /**
     * Return a fully-qualified dataSharingSettings resource name string.
     *
     * @param {string} account
     * @returns {string} Resource name string.
     */
    dataSharingSettingsPath(account) {
        return this.pathTemplates.dataSharingSettingsPathTemplate.render({
            account: account,
        });
    }
    /**
     * Parse the account from DataSharingSettings resource.
     *
     * @param {string} dataSharingSettingsName
     *   A fully-qualified path representing DataSharingSettings resource.
     * @returns {string} A string representing the account.
     */
    matchAccountFromDataSharingSettingsName(dataSharingSettingsName) {
        return this.pathTemplates.dataSharingSettingsPathTemplate.match(dataSharingSettingsName).account;
    }
    /**
     * Return a fully-qualified dataStream resource name string.
     *
     * @param {string} property
     * @param {string} data_stream
     * @returns {string} Resource name string.
     */
    dataStreamPath(property, dataStream) {
        return this.pathTemplates.dataStreamPathTemplate.render({
            property: property,
            data_stream: dataStream,
        });
    }
    /**
     * Parse the property from DataStream resource.
     *
     * @param {string} dataStreamName
     *   A fully-qualified path representing DataStream resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromDataStreamName(dataStreamName) {
        return this.pathTemplates.dataStreamPathTemplate.match(dataStreamName)
            .property;
    }
    /**
     * Parse the data_stream from DataStream resource.
     *
     * @param {string} dataStreamName
     *   A fully-qualified path representing DataStream resource.
     * @returns {string} A string representing the data_stream.
     */
    matchDataStreamFromDataStreamName(dataStreamName) {
        return this.pathTemplates.dataStreamPathTemplate.match(dataStreamName)
            .data_stream;
    }
    /**
     * Return a fully-qualified firebaseLink resource name string.
     *
     * @param {string} property
     * @param {string} firebase_link
     * @returns {string} Resource name string.
     */
    firebaseLinkPath(property, firebaseLink) {
        return this.pathTemplates.firebaseLinkPathTemplate.render({
            property: property,
            firebase_link: firebaseLink,
        });
    }
    /**
     * Parse the property from FirebaseLink resource.
     *
     * @param {string} firebaseLinkName
     *   A fully-qualified path representing FirebaseLink resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromFirebaseLinkName(firebaseLinkName) {
        return this.pathTemplates.firebaseLinkPathTemplate.match(firebaseLinkName)
            .property;
    }
    /**
     * Parse the firebase_link from FirebaseLink resource.
     *
     * @param {string} firebaseLinkName
     *   A fully-qualified path representing FirebaseLink resource.
     * @returns {string} A string representing the firebase_link.
     */
    matchFirebaseLinkFromFirebaseLinkName(firebaseLinkName) {
        return this.pathTemplates.firebaseLinkPathTemplate.match(firebaseLinkName)
            .firebase_link;
    }
    /**
     * Return a fully-qualified googleAdsLink resource name string.
     *
     * @param {string} property
     * @param {string} google_ads_link
     * @returns {string} Resource name string.
     */
    googleAdsLinkPath(property, googleAdsLink) {
        return this.pathTemplates.googleAdsLinkPathTemplate.render({
            property: property,
            google_ads_link: googleAdsLink,
        });
    }
    /**
     * Parse the property from GoogleAdsLink resource.
     *
     * @param {string} googleAdsLinkName
     *   A fully-qualified path representing GoogleAdsLink resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromGoogleAdsLinkName(googleAdsLinkName) {
        return this.pathTemplates.googleAdsLinkPathTemplate.match(googleAdsLinkName)
            .property;
    }
    /**
     * Parse the google_ads_link from GoogleAdsLink resource.
     *
     * @param {string} googleAdsLinkName
     *   A fully-qualified path representing GoogleAdsLink resource.
     * @returns {string} A string representing the google_ads_link.
     */
    matchGoogleAdsLinkFromGoogleAdsLinkName(googleAdsLinkName) {
        return this.pathTemplates.googleAdsLinkPathTemplate.match(googleAdsLinkName)
            .google_ads_link;
    }
    /**
     * Return a fully-qualified keyEvent resource name string.
     *
     * @param {string} property
     * @param {string} key_event
     * @returns {string} Resource name string.
     */
    keyEventPath(property, keyEvent) {
        return this.pathTemplates.keyEventPathTemplate.render({
            property: property,
            key_event: keyEvent,
        });
    }
    /**
     * Parse the property from KeyEvent resource.
     *
     * @param {string} keyEventName
     *   A fully-qualified path representing KeyEvent resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromKeyEventName(keyEventName) {
        return this.pathTemplates.keyEventPathTemplate.match(keyEventName).property;
    }
    /**
     * Parse the key_event from KeyEvent resource.
     *
     * @param {string} keyEventName
     *   A fully-qualified path representing KeyEvent resource.
     * @returns {string} A string representing the key_event.
     */
    matchKeyEventFromKeyEventName(keyEventName) {
        return this.pathTemplates.keyEventPathTemplate.match(keyEventName)
            .key_event;
    }
    /**
     * Return a fully-qualified measurementProtocolSecret resource name string.
     *
     * @param {string} property
     * @param {string} data_stream
     * @param {string} measurement_protocol_secret
     * @returns {string} Resource name string.
     */
    measurementProtocolSecretPath(property, dataStream, measurementProtocolSecret) {
        return this.pathTemplates.measurementProtocolSecretPathTemplate.render({
            property: property,
            data_stream: dataStream,
            measurement_protocol_secret: measurementProtocolSecret,
        });
    }
    /**
     * Parse the property from MeasurementProtocolSecret resource.
     *
     * @param {string} measurementProtocolSecretName
     *   A fully-qualified path representing MeasurementProtocolSecret resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromMeasurementProtocolSecretName(measurementProtocolSecretName) {
        return this.pathTemplates.measurementProtocolSecretPathTemplate.match(measurementProtocolSecretName).property;
    }
    /**
     * Parse the data_stream from MeasurementProtocolSecret resource.
     *
     * @param {string} measurementProtocolSecretName
     *   A fully-qualified path representing MeasurementProtocolSecret resource.
     * @returns {string} A string representing the data_stream.
     */
    matchDataStreamFromMeasurementProtocolSecretName(measurementProtocolSecretName) {
        return this.pathTemplates.measurementProtocolSecretPathTemplate.match(measurementProtocolSecretName).data_stream;
    }
    /**
     * Parse the measurement_protocol_secret from MeasurementProtocolSecret resource.
     *
     * @param {string} measurementProtocolSecretName
     *   A fully-qualified path representing MeasurementProtocolSecret resource.
     * @returns {string} A string representing the measurement_protocol_secret.
     */
    matchMeasurementProtocolSecretFromMeasurementProtocolSecretName(measurementProtocolSecretName) {
        return this.pathTemplates.measurementProtocolSecretPathTemplate.match(measurementProtocolSecretName).measurement_protocol_secret;
    }
    /**
     * Return a fully-qualified property resource name string.
     *
     * @param {string} property
     * @returns {string} Resource name string.
     */
    propertyPath(property) {
        return this.pathTemplates.propertyPathTemplate.render({
            property: property,
        });
    }
    /**
     * Parse the property from Property resource.
     *
     * @param {string} propertyName
     *   A fully-qualified path representing Property resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromPropertyName(propertyName) {
        return this.pathTemplates.propertyPathTemplate.match(propertyName).property;
    }
    /**
     * Terminate the gRPC channel and close the client.
     *
     * The client will no longer be usable and all future behavior is undefined.
     * @returns {Promise} A promise that resolves when the client is closed.
     */
    close() {
        if (this.analyticsAdminServiceStub && !this._terminated) {
            return this.analyticsAdminServiceStub.then(stub => {
                this._terminated = true;
                stub.close();
            });
        }
        return Promise.resolve();
    }
}
exports.AnalyticsAdminServiceClient = AnalyticsAdminServiceClient;
